### ISSUE TITLE
Associated Models/Attachments

### ISSUE THREAD

**Issue by dustmoo** [laserlemon/vestal_versions#9]
First of all, I love the simplicity and DRY of this plugin, fantastic work!

Most of my models, and specifically the ones I would like to track have associated models.

For instance, a client record has an avatar attachment (paperclip),
Many phone numbers, and many email addresses.

I was thinking that we could possibly store the attributes for these associations in the version database and then use a build command to show them, without creating them while iterating through versions.

Paperclip attachments are harder, as if they get destroyed, the files are gone, even if you DO track the file names.

Anyway, I would be happy to collaborate with you on this feature, as I think it is invaluable to true versioning. If you don't already have something like this in the works.

Thanks!

Dustin

**Comment by laserlemon** [comment]
Dustin,

Thanks for the suggestion. I agree that versioning associations would be pretty great. I know there are other versioning plugins/gems that support it. But looking around, I haven't seen a very clean way to handle it.

Building associated models based on attributes stored in the versions table is possible, for instance. But what if the association is autosave? A revert_to! would recreate those associations. It seems to be a pretty large can of worms and I don't know if it's possible to cleanly cover all the scenarios.

And you're right. Versioning a Paperclip attachement is impossible (I think) without somehow overriding Paperclip's file deletion.

I'm considering integration of versioned belongs_to associations into vestal_versions, as seen in my [belongs_to_versioned](http://github.com/laserlemon/belongs_to_versioned) gem, but the other associations may prove too messy.

As always, I'm open to contribution, so if you have an idea how this functionality could be included without making too much of a mess, please fork away! I'd love to see what you have in mind. Thanks again,

Steve

**Comment by dustmoo** [comment]
Steve, Thank you for your thoughtful reply. I can tell you first hand now that this is definitely no simple task. I went ahead and worked <a href="http://github.com/fatjam/acts_as_revisable">acts_as_revisable</a> into my application. And needless to say, I discovered first-hand all the potential problems. But in the end I got it all to work. (Sans the paperclip versioning, which at this point still seems impossible)

Now, Rich's plugin works by creating a Revisable class that basically shares your existing table with your main Model class, and then creates copies in that table. With associations the duplicated objects can get very out of hand and is not DRY. It really isn't a solution I would recommend to others if they want to keep thier versions (I am constantly destroying old ones when changes are accepted in my app to keep the DB clean)

Essentially, what I ran into with the acts_as_revisable gem is exactly what you said, he didn't really think about all the different association scenarios.

In my case, I have a Client object that has_many phone numbers, emails, urls etc. The problem was, that when either of those were updated, created or destroyed I would need to trigger a version of the Client object. But that trigger had to happen "before" the new association was created, modified or destroyed. 

To do this acts_as_revisable relys on the :before_add and :before_remove association methods. But this causes issues:

I was creating the associated objects through virtual attributes in my client model, (using Ryan's technique from Complex Forms in Railscasts). So if I am updating my Client record, I could be adding 3 phone numbers and 2 emails. If each is triggering a revision using the :before_add association method, then I would get 5 versioning calls. (i.e jump from version 1 to 6) There was really no way around this (with his gem) other than doing some manual cleanup of the versions.

In order for this to update cleanly, we would need to create a way for the associated models to communicate what they are doing to the parent model, i.e.  i_am_deleteing, i_am_creating_from_you or something and then write a routine that would basically interpret these messages and then handle the versioning accordingly. 

I would love to give a fork a go at it once my current deadlines are met. In the mean time, perhaps this message can show others that this isn't simply implemented. :D Thanks again for your clean, great gem Steve. I am definitely going to be using it for simple versioning in the future.

**Comment by laserlemon** [comment]
Dustin,

Thanks for all the information. These are the sorts of concerns I need to see if we want to eventually implement versioned associations successfully.

One idea I've had is to extend vestal_versions to also version record deletions (a la acts_as_paranoid). Then it would be possible to conjure up old associated objects without rebuilding them. Food for thought.

I'm looking forward to seeing your work. Thanks for the help!

Steve

**Comment by dustmoo** [comment]
Steve,

That is definitely an idea to chew on. :) Thanks for getting back to me and your time. I look forward to seeing what I can contribute.

Dustin

**Comment by laserlemon** [comment]
A 1.0 release is in the works and association versioning is on the top of the priority list, along with a slew of other handy features.

A quick breakdown of my preliminary plans:
1. Association versioning is twofold: reverting a record to restore _which_ records were associated with it at a given time, **and** reverting those associated records to their individual states at that same given time. These two can be performed independently of each other.
2. The first issue can be approached by reassigning the association ids to the reference the old associations.
3. If it's desired to also restore the associated objects to their historical states, this can be done through simply versioning the associated models as well. A `revert_to` call with an argument representing the effective date and time that the original record was reverted to would restore the associated models to their proper states.

The advantage that vestal_versions has when approaching association versioning is that because reversion doesn't rely on version numbers, reversions can be made deep into the association tree using a datetime, which every model can understand.

But, there are some issues:
1. Restoring the ids of associated models is straightforward for `belongs_to` associations: just version the columns that reference the association. For collection associations (`has_many` and `has_and_belongs_to_many`), we can include a virtual attribute to our versioned changes: `other_ids` in the case of an `others` association. But... what about `has_one` associations? There's no built-in `other_id` method that we can tap into to fool the versioned model.
2. What about associated records that have been deleted? An attempt to revert to a time when they did exist won't magically revive them. I considered extending vestal_versions to include "paranoid" functionality, but I now think this is outside the scope of vestal_versions. Right now, I'm leaning toward accepting that reverting may produce some now non-existent associated records. I'd like to include functionality to allow vestal_versions to play nice with some popular paranoid plugins and properly dig up the lost associated records if they in fact do still exist in AR limbo.
3. To avoid the issue of lost records, we could alternatively store attributes of associated records alongside the original record's versions. But this seems like a lot of bloat to store information that already has a natural home in other database tables.

Long story short: I'm working on it. There are a lot of considerations when trying to version associations and to be honest, I'm unsure if there really is a good, clean way to pull it off. But I'll do my best and I'd appreciate any insight or help that can be given. 

Stay tuned for big additions to vestal_versions!

**Comment by KevinTriplett** [comment]
@dustmoo: I believe that using acts_as_revisable's changeset feature in the parent controller would prevent the multiple parent versions (see Rich's example of a changeset).  Of course, I haven't used aar yet but have been prepping for it and trying to decide which plugin to use.

I'm glad to see that vestal_versions 1.0 may have association versioning, I'm glad I waited to implement versioning in my project. I think using datetime may be a smart way to handle it (there may be some strange race conditions where multiple users are updating the same objects). I also think that revisions and the revisable sharing the same table, like aar does, might have prevented the id collisions, but of course you wouldn't want to make that big a change at this point. :)

I'm not sure that paranoid deletions are out of scope of a versioning plugin, since deleting an association is essentially a version of the parent object. In my app, I want to track the changes to contract line items, including restoring accidentally deleted line items or a customer changing their mind.

As far as DB size goes, I don't believe that extraneous records should be an issue unless your app is dealing with an enormous number of rows, which should be a rare instance under modern hardware.

Good luck on the new release -- looking forward to it!

**Comment by laserlemon** [comment]
Version 1.0 was just released and unfortunately association versioning is not on board. I still haven't landed on the implementation that "feels right." But I didn't want some of the other 1.0 features to be held up any longer.

Without adding a lot of "paranoid" functionality to the plugin (which I think may cross the line into being _too_ magical), vestal_versions will probably not begin to store attributes of deleted, associated objects. Maybe down the road. That being said, the first step will be to add a `revert_to` option to the standard association methods (`belongs_to`, `has_many`, etc.). This would tap into the association proxies and revert the associated objects to a specified version (most likely a date/time) upon loading them. To be clear, no records will be "resurrected," only reverted to a pertinent point in time through the association.

**Comment by kcore** [comment]
when can i expect a update for storing changes to associations "has_many/belongs_to" of a model?
